kafka-connect/src/main/java/io/tabular/iceberg/connect/data/RecordRouter.java

```diff
@@ -0,0 +1,194 @@
+    public void write(SinkRecord record) {
+      // route to all tables
+      tables.forEach(
+          tableName -> {
+            writers.write(tableName, record, false);
+          });
+    }
+  }
+
+  private static class StaticRecordRouter extends RecordRouter {
+    private final String routeField;
+    private final WriterManager writers;
+    private final IcebergSinkConfig config;
+
+    StaticRecordRouter(WriterManager writers, IcebergSinkConfig config) {
+      this.routeField = config.tablesRouteField();
+      this.writers = writers;
+      this.config = config;
+    }
+
+    @Override
+    public void write(SinkRecord record) {
+      String routeValue = extractRouteValue(record.value(), routeField);
+      if (routeValue != null) {
+        config
+            .tables()
+            .forEach(
+                tableName ->
+                    config
+                        .tableConfig(tableName)
+                        .routeRegex()
+                        .ifPresent(
+                            regex -> {
+                              boolean matches;
+                              try {
+                                matches = regex.matcher(routeValue).matches();
+                              } catch (Exception error) {
+                                throw new WriteException.RouteRegexException(error);
+                              }
+                              if (matches) {
+                                writers.write(tableName, record, false);
+                              }
+                            }));
+      }
+    }
+  }
+
+  private static class DynamicRecordRouter extends RecordRouter {
+    private final String routeField;
+    private final WriterManager writers;
+
+    DynamicRecordRouter(WriterManager writers, String routeField) {
+      this.routeField = routeField;
+      this.writers = writers;
+    }
+
+    @Override
+    public void write(SinkRecord record) {
+      String routeValue = extractRouteValue(record.value(), routeField);
+      if (routeValue != null) {
+        String tableName = routeValue.toLowerCase();
+        writers.write(tableName, record, true);
+      }
+    }
+  }
+
+  private static class ErrorHandlingRecordRouter extends RecordRouter {
+    private final WriteExceptionHandler handler;
+    private final WriterManager writers;
+    private final RecordRouter router;
+    private final FailedRecordFactory failedRecordFactory;
+
+    ErrorHandlingRecordRouter(
+        RecordRouter baseRouter,
+        WriteExceptionHandler handler,
+        WriterManager writers,
+        FailedRecordFactory factory) {
+      this.router = baseRouter;
+      this.handler = handler;
+      this.writers = writers;
+      this.failedRecordFactory = factory;
+    }
+
+    @Override
+    public void write(SinkRecord record) {
+      if (failedRecordFactory.isFailedTransformRecord(record)) {
+        writers.write(failedRecordFactory.tableName(record), record, false);
+      } else {
```

Comment:
I still think things could be simpler/cleaner for both us (maintainers) and users if we were to just implement the "third mode" (really dynamic table with fallback, assuming this can be done without breaking changes). 
We would be able to remove this entire `FailedRecordFactory` configuration if we were to do that leaving just one additional config `WriteExceptionHandler`.

Reply:
I can't see this without leaving the possibility to infinite loop Kafka Connect which imho is a deal breaker.

Reply:
Third mode implemented, see RecordRouterTests